### ISSUE TITLE
Create archetype for versioned bundles

### ISSUE THREAD

**Issue by aalexandrov** [stratosphere/peel#42]
Make this aligned [with the command listed in the "Getting Started" guide](http://peel-framework.org/getting-started.html#maven-archetype). 

The archetype should bootstrap [the running example from the manual](http://peel-framework.org/manual/motivation.html#running-example). You can use the [tamara-msc-thesis](https://github.com/aalexandrov/tamara-msc-thesis) bundle as a basis.

**Comment by aalexandrov** [comment]
@carabolic please prepare and push an initial version of a repository called `peel-wordcount-bundle` in GitHub. 

- [] The repository should follow the layout [defined in the "Motivation" section of the website](http://peel-framework.org/manual/motivation.html#running-example). 
- [] The words should be generated by a random word generator shipped as part of the bundle as a Flink program.
- [] The `flink-jobs and `spark-jobs` sub-packages should be built against the latest stable Flink and Spark versions (0.9.0 and 1.4.0).

**Comment by carabolic** [comment]
Considering the flink and spark jobs: should they be implemented using Java 7, Java 8 (lambdas), or Scala?

**Comment by aalexandrov** [comment]
I would (unsurprisingly) suggest Scala...

**Comment by akunft** [comment]
+1 for scala

**Comment by carabolic** [comment]
Okay. So I've added wordcount examples for Apache Flink 0.9.0 and Apache Spark 1.4.0.

I uploaded the current version to https://github.com/carabolic/peel-wordcount-bundle. There is still some work left:
* implementing the word generator
* cleaning up the pom and config files

**Comment by aalexandrov** [comment]
Thanks, you can also remove `peel-wordcount-spark-datagens`.

**Comment by qmlmoon** [comment]
I have implemented a word generator in procrustes, you can have a look if it helps.
  https://github.com/qmlmoon/procrustes/blob/master/procrustes-datagen/src/main/scala/eu/stratosphere/procrustes/datagen/spark/SparkDocumentGenerator.scala

**Comment by aalexandrov** [comment]
@carabolic You need to migrate the generator to Flink. Please, follow the basic pattern used in the [FlinkIntGenerator](https://github.com/carabolic/peel-wordcount-bundle/blob/master/peel-wordcount-flink-datagens/src/main/scala/eu/stratosphere/procrustes/datagen/flink/FlinkIntGenerator.scala#L20):

```scala
    val seed = // set seed
    val dop = coresPerWorker * numberOfWorkers // compute DOP
    val n = dop * tuplesPerTask // compute total cardinality
    val C = 1 // how much seeds are consumed per 'map' call
    val numberOfSeeds = // compute total number 
    val dataset = env
      .fromParallelCollection(new NumberSequenceIterator(1, n))
      .setParallelism(dop)
      .map(i => dictionary.sample(new RanHash(seed + 3*i + 0))))
```

**Comment by carabolic** [comment]
Thanks for giving me directions. I've [implemented](https://github.com/carabolic/peel-wordcount-bundle) the word generator using Apache Flink. I've also added the according XML entries to have an experiments suite that runs wordcount on the generated documents. The word generator itself seems to be working, but I get an `Exception` when I try to run it using Peel (see below).

I think Peel is trying to delete the generated data (`*.hdfs*.rmr`), which for the first run is bound to fail.

Any ideas what went wrong?



```bash
15-07-16 22:43:38 [INFO] Materializing experiment input data sets
15-07-16 22:43:42 [INFO] Generating data for target 'hdfs://localhost:9000/tmp/input/wordcount/random-strings-10000'
15-07-16 22:43:45 [ERROR] Exception for experiment 'wordcount.flink.generated' in suite 'wordcount.default.generated': Unable to remove path in HDFS.
15-07-16 22:43:45 [INFO] Tearing down systems with EXPERIMENT lifespan
15-07-16 22:43:45 [ERROR] Exception in suite 'wordcount.default.generated': Unable to remove path in HDFS.
15-07-16 22:43:45 [INFO] ############################################################
15-07-16 22:43:45 [INFO] Tearing down systems with SUITE lifespan
15-07-16 22:43:46 [INFO] Tearing down system 'flink-0.9.0'
15-07-16 22:43:48 [INFO] Shut down system 'flink-0.9.0'.
15-07-16 22:43:49 [INFO] Tearing down system 'hdfs-2.7.1'
15-07-16 22:44:20 [INFO] Formatting namenode
15-07-16 22:44:28 [INFO] Fixing data directories
15-07-16 22:44:28 [INFO] Initializing data directory /tmp/hdfs-2/data at datanode localhost
15-07-16 22:44:29 [INFO] Shut down system 'hdfs-2.7.1'.
15-07-16 22:44:29 [INFO] System 'hdfs-2.4.1' is already down
Unexpected error: Unable to remove path in HDFS.
java.lang.RuntimeException: Unable to remove path in HDFS.
	at eu.stratosphere.peel.core.util.shell$.$bang(shell.scala:57)
	at eu.stratosphere.peel.extensions.hadoop.beans.system.HDFSFileSystem$class.rmr(HDFSFileSystem.scala:23)
	at eu.stratosphere.peel.extensions.hadoop.beans.system.HDFS2.rmr(HDFS2.scala:21)
	at eu.stratosphere.peel.core.cli.command.suite.Run$$anonfun$run$2$$anonfun$apply$8.apply(Run.scala:98)
	at eu.stratosphere.peel.core.cli.command.suite.Run$$anonfun$run$2$$anonfun$apply$8.apply(Run.scala:97)
	at scala.collection.immutable.Set$Set1.foreach(Set.scala:74)
	at eu.stratosphere.peel.core.cli.command.suite.Run$$anonfun$run$2.apply(Run.scala:97)
	at eu.stratosphere.peel.core.cli.command.suite.Run$$anonfun$run$2.apply(Run.scala:73)
	at scala.collection.immutable.List.foreach(List.scala:318)
	at eu.stratosphere.peel.core.cli.command.suite.Run.run(Run.scala:73)
	at eu.stratosphere.peel.core.cli.Peel$.main(Peel.scala:78)
	at eu.stratosphere.peel.core.cli.Peel.main(Peel.scala)
15-07-16 22:44:29 [INFO] Closing org.springframework.context.support.ClassPathXmlApplicationContext@4f51b3e0: startup date [Thu Jul 16 22:42:41 CEST 2015]; root of context hierarchy
```

**Comment by qmlmoon** [comment]
I tried your code. This error message is actually a bit annoying. 
Peel tries to run the data generator, but the command can't be parsed. Then peel tries to clean the output directory but there's no such directory.
To fix this, simply change ${datagen.tuples.per.task} --> ${datagen.tuples.per.tasks}  in experiment.wordcount.xml file line 13.
A small mistake, which is hard to debug:-)

**Comment by qmlmoon** [comment]
And you ref both hdfs-2.7.1 and hdfs-2.4.1, I think we should unify them.

**Comment by aalexandrov** [comment]
I'm fixing this [here](https://github.com/stratosphere/peel-wordcount-bundle).

**Comment by aalexandrov** [comment]
@qmlmoon we need to react earlier on parsing errors. Can you open an issue about that?

**Comment by aalexandrov** [comment]
@carabolic the code at the moment looks like:

```scala
    val randomWords = env
      // create a sequence [1 .. N] to create N words
      .fromParallelCollection(new NumberSequenceIterator(1, n))
      // set up workers
      .setParallelism(dop)
      // map every n <- [1 .. N] to a random character string
      .map(i => generateWord(seed + (MAX_LENGTH + 1) * i))
```

where `generateWord` is

```scala
  def generateWord(seed : Long): String = {
    val ran = new RanHash(seed)
    val length = ran.nextInt(MAX_LENGTH)
    val strBld = new StringBuilder(length)
    for (i <- 0 until length) {
      val c = ('a'.toInt + ran.nextInt(NUM_CHARACTERS)).toChar
      strBld.append(c)
    }
    return strBld.mkString
  }
```

This will generate N different words. What you need to do is to 

* have a generator function ( `word(i: Int): String`, similar to `generateWord`) that generates K different words (where K is the size of the dictionary and `i` is the position of the word in the generated sequence), and then pick words 
* in the mapper, pick the index `i` at random, based on a given distribution (if you look at the code in the other projects you will find abstractions over distributions like Uniform, Pareto, Gaussian), and the feed it to the word generator, e.g. `map(i => word(/* sample index i */)`;

**Comment by lauritzthamsen** [comment]
generating projects from the archetypes currently creates an empty folder `null` in the bundle-src

**Comment by aalexandrov** [comment]
@lauritzthamsen can you post a shell trace?

**Comment by lauritzthamsen** [comment]
yes, here is what i see when i generate from the archetype: 

> ~...ries/peel-bundles ✔ export BUNDLE_AID=test                       # bundle name
~...ries/peel-bundles ✔ export BUNDLE_PKG=de.tuberlin.cit.experiments       # bundle package
~...ries/peel-bundles ✔ export BUNDLE_GID=de.tuberlin.cit                     # bundle group
~...ries/peel-bundles ✔ mvn archetype:generate -B                         \
>     -Dpackage="$BUNDLE_PKG"                       \
>     -DgroupId="$BUNDLE_GID"                       \
>     -DartifactId="$BUNDLE_AID"                    \
>     -DarchetypeGroupId=org.peelframework          \
>     -DarchetypeArtifactId=peel-flinkspark-bundle  \
>     -DarchetypeVersion=1.0-SNAPSHOT
[INFO] Scanning for projects...
[INFO]                                                                         
[INFO] ------------------------------------------------------------------------
[INFO] Building Maven Stub Project (No POM) 1
[INFO] ------------------------------------------------------------------------
[INFO] 
[INFO] >>> maven-archetype-plugin:2.3:generate (default-cli) > generate-sources @ standalone-pom >>>
[INFO] 
[INFO] <<< maven-archetype-plugin:2.3:generate (default-cli) < generate-sources @ standalone-pom <<<
[INFO] 
[INFO] --- maven-archetype-plugin:2.3:generate (default-cli) @ standalone-pom ---
[INFO] Generating project in Batch mode
[INFO] Archetype repository not defined. Using the one from [org.peelframework:peel-flinkspark-bundle:1.0.0-beta] found in catalog remote
[INFO] ----------------------------------------------------------------------------
[INFO] Using following parameters for creating project from Archetype: peel-flinkspark-bundle:1.0-SNAPSHOT
[INFO] ----------------------------------------------------------------------------
[INFO] Parameter: groupId, Value: de.tuberlin.cit
[INFO] Parameter: artifactId, Value: test
[INFO] Parameter: version, Value: 1.0-SNAPSHOT
[INFO] Parameter: package, Value: de.tuberlin.cit.experiments
[INFO] Parameter: packageInPathFormat, Value: de/tuberlin/cit/experiments
[INFO] Parameter: version, Value: 1.0-SNAPSHOT
[INFO] Parameter: package, Value: de.tuberlin.cit.experiments
[INFO] Parameter: groupId, Value: de.tuberlin.cit
[INFO] Parameter: artifactId, Value: test
[INFO] Parent element not overwritten in /Users/lauritz/Code/Repositories/peel-bundles/test/test-bundle/pom.xml
[INFO] Parent element not overwritten in /Users/lauritz/Code/Repositories/peel-bundles/test/test-peelextensions/pom.xml
[INFO] Parent element not overwritten in /Users/lauritz/Code/Repositories/peel-bundles/test/test-datagens/pom.xml
[INFO] Parent element not overwritten in /Users/lauritz/Code/Repositories/peel-bundles/test/test-flink-jobs/pom.xml
[INFO] Parent element not overwritten in /Users/lauritz/Code/Repositories/peel-bundles/test/test-spark-jobs/pom.xml
[INFO] project created from Archetype in dir: /Users/lauritz/Code/Repositories/peel-bundles/test
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 3.121 s
[INFO] Finished at: 2015-08-17T14:46:08+02:00
[INFO] Final Memory: 13M/198M
[INFO] ------------------------------------------------------------------------
~...ries/peel-bundles ✔ cd 
adaptive-iterations/ test/                
~...ries/peel-bundles ✔ cd test/
null/                pom.xml              test-bundle/         test-datagens/       test-flink-jobs/     test-peelextensions/ test-spark-jobs/     
~...ries/peel-bundles ✔ ll test/
total 16
drwxr-xr-x  2 lauritz  staff    68 Aug 17 14:46 null
-rw-r--r--  1 lauritz  staff  7659 Aug 17 14:46 pom.xml
drwxr-xr-x  4 lauritz  staff   136 Aug 17 14:46 test-bundle
drwxr-xr-x  4 lauritz  staff   136 Aug 17 14:46 test-datagens
drwxr-xr-x  4 lauritz  staff   136 Aug 17 14:46 test-flink-jobs
drwxr-xr-x  4 lauritz  staff   136 Aug 17 14:46 test-peelextensions
drwxr-xr-x  4 lauritz  staff   136 Aug 17 14:46 test-spark-jobs

**Comment by aalexandrov** [comment]
@lauritzthamsen should be fixed in the latest snapshot.

**Comment by aalexandrov** [comment]
Resolved with 58bdebd6b0fded02fb08c156961d91fae1a06f85.
